File: Day26/Part3/temporal_switch.py
from collections import deque
from typing import List

def min_presses(n: int, m: int, connections: List[List[int]]) -> int:
    masks = [0]*n
    for i, lamps in enumerate(connections):
        mask = 0
        for lamp in lamps:
            mask ^= (1 << lamp)
        masks[i] = mask
    target = (1 << m) - 1
    q = deque([0])
    dist = {0: 0}
    while q:
        s = q.popleft()
        if s == target:
            return dist[s]
        d = dist[s]
        for i in range(n):
            ns = s ^ masks[i]
            if ns not in dist:
                dist[ns] = d + 1
                q.append(ns)
    return -1

if __name__ == "__main__":
    n = 3; m = 3
    connections = [
        [0,1],
        [1,2],
        [0,2],
    ]
    print(min_presses(n, m, connections))  # Expected 3

File: Day26/Part3/temporal_switch.cpp
#include <bits/stdc++.h>
using namespace std;

int min_presses(int n, int m, const vector<vector<int>>& connections){
    vector<int> masks(n, 0);
    for(int i=0;i<n;++i){
        int mask = 0;
        for(int lamp: connections[i]) mask ^= (1 << lamp);
        masks[i] = mask;
    }
    int target = (1<<m) - 1;
    queue<int> q; q.push(0);
    vector<int> dist(1<<m, -1); dist[0] = 0;
    while(!q.empty()){
        int s = q.front(); q.pop();
        if(s == target) return dist[s];
        for(int i=0;i<n;++i){
            int ns = s ^ masks[i];
            if(dist[ns] == -1){
                dist[ns] = dist[s] + 1;
                q.push(ns);
            }
        }
    }
    return -1;
}

int main(){
    int n=3, m=3;
    vector<vector<int>> connections = {
        {0,1}, {1,2}, {0,2}
    };
    cout << min_presses(n, m, connections) << "\n"; // Expected 3
    return 0;
}
